pkg/controller/podupdater/util.go

```diff
@@ -0,0 +1,181 @@
+	"k8s.io/klog/v2"
+
+	k8sutil "github.com/openyurtio/openyurt/pkg/controller/podupdater/kubernetes"
+)
+
+// GetDaemonsetPods get all pods belong to the given daemonset
+func GetDaemonsetPods(podLister corelisters.PodLister, ds *appsv1.DaemonSet) ([]*corev1.Pod, error) {
+	dsPods := make([]*corev1.Pod, 0)
+	dsPodsNames := make([]string, 0)
+	pods, err := podLister.Pods(ds.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	for i, pod := range pods {
+		owner := metav1.GetControllerOf(pod)
+		if owner == nil {
+			continue
+		}
+		if owner.UID == ds.UID {
+			dsPods = append(dsPods, pods[i])
+			dsPodsNames = append(dsPodsNames, pod.Name)
+		}
+	}
+
+	if len(dsPods) > 0 {
+		klog.V(4).Infof("Daemonset %v has pods %v", ds.Name, dsPodsNames)
+	}
+	return dsPods, nil
+}
+
+// IsDaemonsetPodLatest check whether pod is latest by comparing its Spec with daemonset's
+// If pod is latest, return true, otherwise return false
+func IsDaemonsetPodLatest(ds *appsv1.DaemonSet, pod *corev1.Pod) bool {
+	hash := k8sutil.ComputeHash(&ds.Spec.Template, ds.Status.CollisionCount)
+	klog.V(4).Infof("compute hash: %v", hash)
+	generation, err := GetTemplateGeneration(ds)
+	if err != nil {
+		generation = nil
+	}
+
+	klog.V(5).Infof("daemonset %v revision hash is %v", ds.Name, hash)
+	klog.V(5).Infof("daemonset %v generation is %v", ds.Name, generation)
+
+	templateMatches := generation != nil && pod.Labels[extensions.DaemonSetTemplateGenerationKey] == fmt.Sprint(generation)
+	hashMatches := len(hash) > 0 && pod.Labels[extensions.DefaultDaemonSetUniqueLabelKey] == hash
+	return hashMatches || templateMatches
+}
+
+// GetTemplateGeneration get annotation "deprecated.daemonset.template.generation" of the given daemonset
+func GetTemplateGeneration(ds *appsv1.DaemonSet) (*int64, error) {
+	annotation, found := ds.Annotations[appsv1.DeprecatedTemplateGeneration]
+	if !found {
+		return nil, nil
+	}
+	generation, err := strconv.ParseInt(annotation, 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	return &generation, nil
+}
+
+// NodeReadyByName check if the given node is ready
+func NodeReadyByName(nodeList corelisters.NodeLister, nodeName string) (bool, error) {
+	node, err := nodeList.Get(nodeName)
+	if err != nil {
+		return false, err
+	}
+
+	return NodeReady(&node.Status), nil
+}
+
+// NodeReady check if the given node status is ready
+func NodeReady(nodeStatus *corev1.NodeStatus) bool {
+	for _, cond := range nodeStatus.Conditions {
+		if cond.Type == corev1.NodeReady {
+			return cond.Status == corev1.ConditionTrue
+		}
+	}
+	return false
+}
+
+// SetPodUpgradeAnnotation calculate and set annotation "apps.openyurt.io/pod-upgradable" to pod
+func SetPodUpgradeAnnotation(clientset client.Interface, ds *appsv1.DaemonSet, pod *corev1.Pod) error {
+	ok := IsDaemonsetPodLatest(ds, pod)
+
+	var res bool
+	if !ok {
+		res = true
+	}
+
+	metav1.SetMetaDataAnnotation(&pod.ObjectMeta, PodUpgradableAnnotation, strconv.FormatBool(res))
```

Comment:
please delete line 117~120, and rename `ok` to `isUpgraded`, then modify this line as following:
> metav1.SetMetaDataAnnotation(&pod.ObjectMeta, PodUpgradableAnnotation, strconv.FormatBool(!isUpgraded))